File: frontend/src/services/security-lgpd/consentService.ts
/**
 * Consent Management Service
 * Gestão de Consentimentos LGPD (Art. 7, 8, 9)
 *
 * Serviço para gerenciamento completo de consentimentos:
 * - Registro de consentimento
 * - Consulta de consentimentos por titular
 * - Revogação de consentimento
 * - Listagem de finalidades
 * - Listagem de bases legais
 */

import { registerConsent as registerConsentApi, getConsents as getConsentsApi, revokeConsent as revokeConsentApi, listPurposes as listPurposesApi, listLegalBases as listLegalBasesApi } from '@/types/generated/security-lgpd/lgpd-consentimento/lgpd-consentimento';
import type {
  ConsentRequest,
  RevokeConsentParams,
  StandardResponse,
} from '@/types/generated/security-lgpd/conectaPROLGPDSecurityAPI.schemas';


/**
 * Service para gestão de consentimentos LGPD
 */
export class ConsentService {
  /**
   * Registra novo consentimento do titular
   * Conforme Art. 7 da LGPD
   */
  static async registerConsent(
    request: ConsentRequest
  ): Promise<StandardResponse> {
    return registerConsentApi(request);
  }

  /**
   * Consulta todos os consentimentos de um titular
   * @param titularId UUID do titular dos dados
   */
  static async getConsents(
    titularId: string
  ): Promise<StandardResponse> {
    return getConsentsApi(titularId);
  }

  /**
   * Revoga um consentimento específico
   * @param consentId ID do consentimento
   * @param reason Motivo da revogação (mínimo 5 caracteres)
   */
  static async revokeConsent(
    consentId: string,
    reason: string
  ): Promise<StandardResponse> {
    const params: RevokeConsentParams = { reason };
    return revokeConsentApi(consentId, params);
  }

  /**
   * Lista finalidades de consentimento disponíveis
   */
  static async listPurposes(): Promise<StandardResponse> {
    return listPurposesApi();
  }

  /**
   * Lista bases legais LGPD disponíveis
   * Art. 7 a 11 da LGPD
   */
  static async listLegalBases(): Promise<StandardResponse> {
    return listLegalBasesApi();
  }

  /**
   * Valida se um consentimento está ativo
   * Helper para verificação rápida no frontend
   */
  static isConsentActive(consent: any): boolean {
    if (!consent) return false;

    const now = new Date();
    const expiresAt = consent.expires_at ? new Date(consent.expires_at) : null;

    return (
      consent.active === true &&
      consent.revoked === false &&
      (!expiresAt || expiresAt > now)
    );
  }

  /**
   * Calcula dias restantes até expiração
   */
  static getDaysUntilExpiration(expiresAt: string): number {
    const expires = new Date(expiresAt);
    const now = new Date();
    const diff = expires.getTime() - now.getTime();
    return Math.ceil(diff / (1000 * 60 * 60 * 24));
  }

  /**
   * Formata base legal para exibição
   */
  static formatLegalBasis(basis: string): string {
    const formats: Record<string, string> = {
      'consent': 'Consentimento do Titular (Art. 7, I)',
      'legal_obligation': 'Cumprimento de Obrigação Legal (Art. 7, II)',
      'public_administration': 'Administração Pública (Art. 7, III)',
      'research': 'Estudos por Órgão de Pesquisa (Art. 7, IV)',
      'contract': 'Execução de Contrato (Art. 7, V)',
      'judicial_process': 'Processo Judicial (Art. 7, VI)',
      'life_protection': 'Proteção da Vida (Art. 7, VII)',
      'health_protection': 'Proteção à Saúde (Art. 7, VIII)',
      'legitimate_interest': 'Interesse Legítimo (Art. 7, IX)',
      'credit_protection': 'Proteção ao Crédito (Art. 7, X)',
    };

    return formats[basis] || basis;
  }

  /**
   * Formata finalidade para exibição
   */
  static formatPurpose(purpose: string): string {
    const formats: Record<string, string> = {
      'user_registration': 'Cadastro de Usuário',
      'service_provision': 'Prestação de Serviços',
      'marketing': 'Marketing e Publicidade',
      'analytics': 'Análise e Estatísticas',
      'security': 'Segurança da Informação',
      'legal_compliance': 'Compliance Legal',
      'communication': 'Comunicação com o Titular',
      'contract_execution': 'Execução de Contrato',
      'operational': 'Operações do Sistema',
    };

    return formats[purpose] || purpose;
  }
}

export default ConsentService;

File: backend/modules/security_lgpd/services/consent_service.py
"""
Service de Consentimento LGPD.
"""

import logging
import uuid
from datetime import datetime, timedelta
from typing import Any

logger = logging.getLogger(__name__)


class ConsentService:
    """Service para gerenciamento de consentimentos LGPD.

    Encapsula a logica de registro, consulta e revogacao
    de consentimentos de titulares de dados conforme Art. 7 LGPD.
    """

    # Armazenamento em memoria (em producao, usar banco de dados)
    _consents: dict[str, dict[str, Any]] = {}

    def __init__(self):
        """Inicializa o service."""
        pass

    def register_consent(
        self,
        titular_id: str,
        titular_email: str,
        purpose: str,
        legal_basis: str,
        description: str,
        expiration_days: int = 365,
    ) -> dict[str, Any]:
        """Registra consentimento de titular.

        Args:
            titular_id: UUID do titular.
            titular_email: Email do titular.
            purpose: Finalidade do consentimento.
            legal_basis: Base legal LGPD.
            description: Descricao detalhada.
            expiration_days: Dias ate expirar.

        Returns:
            Dict com dados do consentimento registrado.
        """
        consent_id = str(uuid.uuid4())
        now = datetime.utcnow()
        expiration = now + timedelta(days=expiration_days)

        consent_data = {
            "consent_id": consent_id,
            "titular_id": titular_id,
            "titular_email": titular_email,
            "purpose": purpose,
            "legal_basis": legal_basis,
            "description": description,
            "status": "active",
            "created_at": now.isoformat(),
            "expires_at": expiration.isoformat(),
            "revoked_at": None,
            "revoke_reason": None,
        }

        self._consents[consent_id] = consent_data

        logger.info(
            "Consentimento registrado: titular=%s, finalidade=%s",
            titular_id,
            purpose,
        )

        return {
            "consent_id": consent_id,
            "titular_id": titular_id,
            "purpose": purpose,
            "legal_basis": legal_basis,
            "status": "active",
            "expires_at": expiration.isoformat(),
        }

    def get_consents_by_titular(self, titular_id: str) -> dict[str, Any]:
        """Consulta consentimentos de um titular.

        Args:
            titular_id: UUID do titular.

        Returns:
            Dict com lista de consentimentos.
        """
        consents = [c for c in self._consents.values() if c["titular_id"] == titular_id]

        return {
            "titular_id": titular_id,
            "consents": consents,
            "total": len(consents),
        }

    def revoke_consent(self, consent_id: str, reason: str) -> dict[str, Any]:
        """Revoga um consentimento.

        Args:
            consent_id: ID do consentimento.
            reason: Motivo da revogacao.

        Returns:
            Dict com confirmacao da revogacao.

        Raises:
            ValueError: Se consentimento nao encontrado.
        """
        if consent_id not in self._consents:
            raise ValueError(f"Consentimento nao encontrado: {consent_id}")

        consent = self._consents[consent_id]
        consent["status"] = "revoked"
        consent["revoked_at"] = datetime.utcnow().isoformat()
        consent["revoke_reason"] = reason

        logger.info("Consentimento revogado: %s", consent_id)

        return {
            "consent_id": consent_id,
            "status": "revoked",
            "reason": reason,
            "revoked_at": consent["revoked_at"],
        }

    def list_purposes(self) -> list[dict[str, str]]:
        """Lista finalidades de consentimento disponiveis.

        Returns:
            Lista de finalidades conforme Art. 7 LGPD.
        """
        return [
            {"id": "marketing", "description": "Marketing e comunicacoes"},
            {"id": "analytics", "description": "Analise de dados e metricas"},
            {"id": "personalization", "description": "Personalizacao de conteudo"},
            {"id": "service_provision", "description": "Prestacao de servicos"},
            {"id": "legal_obligation", "description": "Obrigacao legal"},
            {"id": "vital_interest", "description": "Interesse vital"},
            {"id": "public_interest", "description": "Interesse publico"},
            {"id": "legitimate_interest", "description": "Interesse legitimo"},
        ]

    def list_legal_bases(self) -> list[dict[str, str]]:
        """Lista bases legais LGPD disponiveis.

        Returns:
            Lista de bases legais conforme Art. 7 LGPD.
        """
        return [
            {"id": "consent", "description": "Consentimento do titular", "article": "Art. 7, I"},
            {"id": "contract", "description": "Execucao de contrato", "article": "Art. 7, V"},
            {"id": "legal_obligation", "description": "Obrigacao legal", "article": "Art. 7, II"},
            {"id": "vital_interest", "description": "Protecao da vida", "article": "Art. 7, VII"},
            {"id": "public_policy", "description": "Politica publica", "article": "Art. 7, III"},
            {"id": "research", "description": "Pesquisa", "article": "Art. 7, IV"},
            {"id": "legitimate_interest", "description": "Interesse legitimo", "article": "Art. 7, IX"},
            {"id": "credit_protection", "description": "Protecao ao credito", "article": "Art. 7, X"},
        ]
